src/rapid_pro_to_engagement_db/rapid_pro_to_engagement_db.py

```diff
@@ -0,0 +1,155 @@
+from core_data_modules.cleaners import URNCleaner
+from core_data_modules.logging import Logger
+from engagement_database.data_models import Message, MessageDirections, MessageStatuses, HistoryEntryOrigin
+from rapid_pro_tools.rapid_pro_client import RapidProClient
+from storage.google_cloud import google_cloud_utils
+
+log = Logger(__name__)
+
+
+def init_rapid_pro_client(google_cloud_credentials_file_path, rapid_pro_domain, rapid_pro_token_file_url):
+    """
+    Initialises a RapidProClient from a Rapid Pro domain and token file url.
+
+    :param google_cloud_credentials_file_path: Path to the google cloud credentials file to use to access the
+                                               rapid pro token.
+    :type google_cloud_credentials_file_path: str
+    :param rapid_pro_domain: Domain which Rapid Pro is running on.
+    :type rapid_pro_domain: str
+    :param rapid_pro_token_file_url: GS url to the Rapid Pro token.
+    :type rapid_pro_token_file_url: str
+    :return: RapidProClient for
+    :rtype: rapid_pro_tools.rapid_pro_client.RapidProClient
+    """
+    log.info(f"Initialising Rapid Pro client for domain {rapid_pro_domain} and auth "
+             f"url {rapid_pro_token_file_url}...")
+    rapid_pro_token = google_cloud_utils.download_blob_to_string(
+        google_cloud_credentials_file_path, rapid_pro_token_file_url).strip()
+    rapid_pro_client = RapidProClient(rapid_pro_domain, rapid_pro_token)
+    log.info("Initialised Rapid Pro client")
+
+    return rapid_pro_client
+
+
+def _engagement_db_has_message(engagement_db, message):
+    """
+    Checks if an engagement database contains a message with the same text, timestamp, and participant_uuid as the
+    given message.
+
+    :param engagement_db: Engagement database to check for the message.
+    :type engagement_db: engagement_database.EngagementDatabase
+    :param message: Message to check for existence.
+    :type message: engagement_database.data_models.Message
+    :return: Whether a message with this text, timestamp, and participant_uuid exists in the engagement database.
+    :rtype: bool
+    """
+    # TODO: Rapid Pro has a bug where timestamps occasionally drift by 1us when runs are archived.
+    #       Confirm this is resolved before entering production with an '==' check on timestamps.
+    filter = lambda q: q \
+        .where("text", "==", message.text) \
+        .where("timestamp", "==", message.timestamp) \
+        .where("participant_uuid", "==", message.participant_uuid)
```

Comment:
Let us also add the dataset name to narrow down the query scope.

Reply:
We can't add the dataset to the search because WS-correction means the dataset can change. As far as the sync from rapid pro is concerned, the dataset is just a best initial guess for how the rest of the pipeline should treat the message, but this can be changed later.

Reply:
I think we're in the same situation - If we add a project name (which can alternatively be encoded in the dataset e.g. CSAP_age), we create the possibility of adding the message multiple times if we change project.